infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java

```diff
@@ -46,58 +56,137 @@
+              + " effect.");
+    }
+
+    // right now allowUserDefinedNamespaces can't be true, but eventually we will implement it.
     if (isNullOrEmpty(defaultNamespaceName) && !allowUserDefinedNamespaces) {
       throw new ConfigurationException(
           "che.infra.kubernetes.namespace.default or "
               + "che.infra.kubernetes.namespace.allow_user_defined must be configured");
     }
   }
 
-  private boolean hasNoPlaceholders(String namespaceName) {
-    return NAMESPACE_NAME_PLACEHOLDERS.keySet().stream().noneMatch(namespaceName::contains);
+  private boolean hasPlaceholders(String namespaceName) {
+    return namespaceName != null
+        && NAMESPACE_NAME_PLACEHOLDERS.keySet().stream().anyMatch(namespaceName::contains);
   }
 
   /**
-   * True if namespace is predefined for all workspaces. False if each workspace will be provided
-   * with a new namespace or provided for each user when using placeholders.
+   * True if namespace is potentially created for the workspace, false otherwise.
+   *
+   * <p>The logic is a little bit non-trivial and best expressed by just fully evaluting the truth
+   * table as below ({@code ...namespace} stands for the legacy namespace property, {@code
+   * ...namespace.default} stands for the namespace default property):
+   *
+   * <pre>{@code
+   * ...namespace    | ...namespace exists | ...namespace.default | creating?
+   * no-placeholders |       no            |       null           | error
+   * no-placeholders |       no            |   no-placeholders    | no
+   * no-placeholders |       no            |    placeholders      | yes
+   * no-placeholders |      yes            |       null           | no
+   * no-placeholders |      yes            |   no-placeholders    | no
+   * no-placeholders |      yes            |    placeholders      | no
+   *  placeholders   |       no            |        null          | error
+   *  placeholders   |       no            |   no-placeholders    | no
+   *  placeholders   |       no            |    placeholders      | yes
+   *  placeholders   |      yes            |        null          | yes
+   *  placeholders   |      yes            |   no-placeholders    | yes
+   *  placeholders   |      yes            |    placeholders      | yes
+   * }</pre>
    */
-  public boolean isPredefined() {
-    return isPredefined;
+  protected boolean isCreatingNamespace(String workspaceId) throws InfrastructureException {
+    boolean legacyWithPlaceholders =
+        isNullOrEmpty(legacyNamespaceName) || hasPlaceholders(legacyNamespaceName);
+    boolean legacyExists =
+        checkNamespaceExists(
+            resolveLegacyNamespaceName(EnvironmentContext.getCurrent().getSubject(), workspaceId));
+
+    if (isNullOrEmpty(defaultNamespaceName) && !legacyExists) {
+      throw new InfrastructureException(
+          "Cannot determine whether a new namespace and service account should be"
```

Comment:
I'm a bit confused.
The message says that `Cannot determine whether a new namespace should be created` but seems it's already created just before invoking `isCreatingNamespace` method:
```java
  public KubernetesNamespace create(String workspaceId) throws InfrastructureException {
 KubernetesNamespace namespace = doCreateNamespace(workspaceId, namespaceName);
    namespace.prepare();

    if (isCreatingNamespace(workspaceId) && !isNullOrEmpty(serviceAccountName)) {
```
```java
  void prepare() throws InfrastructureException {
    KubernetesClient client = clientFactory.create(workspaceId);
    if (get(name, client) == null) {
      create(name, client);
    }
  }
```
Maybe some refactoring can make thinkgs simpler. I'll try to propose something.

Reply:
I don't agree with this. My implementation was meant to be as minimal as possible, injecting only a minimal amount of change needed to support the new properties. During the course of what we originally perceived as a relatively simple hack we uncovered other warts like the `isPredefined()` which I've split apart based on the different circumstances it was being used in. Which in turn uncovered this illogical implementation that was there from the very beginning.

I'm not saying we shouldn't be fixing this, but let's do that outside of this issue. Let's implement the minimal thing here and leave the fixing and proper implementation of this mess to another issue.

Reply:
After all, the very next issue I already started working on - #14908 - is going to shake this up considerably.